### PR TITLE
ARTEMIS-1151 Adapting TimedBuffer and NIO Buffer Pooling

### PR DESCRIPTION
**Pull request by franz1981**
This is the list of improvements:
- NIO/ASYNCIO new TimedBuffer with adapting batch window heuristic
- NIO/ASYNCIO improved TimedBuffer write monitoring with lightweight concurrent performance counters
- NIO/ASYNCIO journal/paging operations benefit from less buffer copy
- NIO/ASYNCIO any buffer copy is always performed with raw batch copy using SIMD instrinsics (System::arrayCopy) or memcpy under the hood
- NIO improved clear buffers using SIMD instrinsics (Arrays::fill) and/or memset
- NIO journal operation perform by default TLABs allocation pooling (off heap) retaining only the last max sized buffer
- NIO improved file copy operations using zero-copy FileChannel::transfertTo
- NIO improved zeroing using pooled single OS page buffer to clean the file + pwrite (on Linux)
- NIO deterministic release of unpooled direct buffers to avoid OOM errors due to slow GC

### CODE REVIEW

File: artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/nio/NIOSequentialFile.java [hunk 1]
```diff
@@ -293,14 +296,49 @@ private void internalWrite(final ByteBuffer bytes,
...
+         //release it to recycle the write buffer if big enough
+         this.factory.releaseBuffer(bytes);
       }
+   }
 
-      if (callback != null) {
-         callback.done();
+   @Override
+   public void copyTo(SequentialFile dstFile) throws IOException {
+      ActiveMQJournalLogger.LOGGER.debug("Copying " + this + " as " + dstFile);
```
Comment (clebertsuconic) [inline artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/nio/NIOSequentialFile.java]: If you make a change here make it:
```
   if (logger.isDebugEnabled()) {
     logger.debug(.....);
  }
```

File: tests/unit-tests/src/test/java/org/apache/activemq/artemis/tests/unit/core/journal/impl/TimedBufferTest.java [hunk 2]
```diff
@@ -205,162 +203,4 @@ public int getRemainingBytes() {
...
 
    }
-
-   /**
-    * This test will verify if the system will switch to spin case the system can't perform sleeps timely
-    * due to proper kernel installations
-    *
-    * @throws Exception
-    */
-   @Test
```
Comment (clebertsuconic) [inline tests/unit-tests/src/test/java/org/apache/activemq/artemis/tests/unit/core/journal/impl/TimedBufferTest.java]: why did you remove this test? it was important when I added it.


I would rather keep a method public and a test than remove a test to make a method private.

Comment (franz1981) [inline tests/unit-tests/src/test/java/org/apache/activemq/artemis/tests/unit/core/journal/impl/TimedBufferTest.java]: Because the new implementation rely on LockSupport.parkNanos to wait (is a MONOTONIC clock on linux AFAIK) and it doesn't wait never the whole time until the given expiration.
In addition the background flusher use the Semaphore only after a "long" period of no writes.
The only practical reason that could lead it to not be triggered in time is dependent by the writers: if a writers wait a lot to push data, the time could expire, but I think is normal no?
